metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -0,0 +1,243 @@
+		options.Node = kubernetes.DiscoverKubernetesNode(config.Host, config.InCluster, client)
+	}
+
+	return kubernetes.NewWatcher(client, resource, options)
+}
+
+// NewResourceMetadataEnricher returns a Enricher configured for kubernetes resource events
+func NewResourceMetadataEnricher(
+	base mb.BaseMetricSet,
+	resource kubernetes.Resource,
+	nodeScope bool) Enricher {
+
+	watcher, err := GetWatcher(base, resource, nodeScope)
+	if err != nil {
+		logp.Warn("Error initializing Kubernetes metadata enricher: %s", err)
+		return nilEnricher
+	}
+
+	if watcher == nil {
+		logp.Info("Kubernetes metricset enriching is disabled")
+		return nilEnricher
+	}
+
+	metaConfig := kubernetes.MetaGeneratorConfig{}
+	if err := base.Module().UnpackConfig(&metaConfig); err != nil {
+		logp.Warn("Error initializing Kubernetes metadata enricher: %s", err)
+		return nilEnricher
+	}
+
+	metaGen := kubernetes.NewMetaGeneratorFromConfig(&metaConfig)
+	enricher := buildMetadataEnricher(watcher,
+		// update
+		func(m map[string]common.MapStr, r kubernetes.Resource) {
+			id := join(r.GetMetadata().GetNamespace(), r.GetMetadata().GetName())
+			m[id] = metaGen.ResourceMetadata(r)
+		},
+		// delete
+		func(m map[string]common.MapStr, r kubernetes.Resource) {
+			id := join(r.GetMetadata().GetNamespace(), r.GetMetadata().GetName())
+			delete(m, id)
+		},
+		// index
+		func(e common.MapStr) string {
+			return join(getString(e, mb.ModuleDataKey+".namespace"), getString(e, "name"))
+		},
+	)
+
+	return enricher
+}
+
+// NewContainerMetadataEnricher returns a Enricher configured for container events
+func NewContainerMetadataEnricher(
+	base mb.BaseMetricSet,
+	nodeScope bool) Enricher {
+
+	watcher, err := GetWatcher(base, &kubernetes.Pod{}, nodeScope)
+	if err != nil {
+		logp.Warn("Error initializing Kubernetes metadata enricher: %s", err)
+		return nilEnricher
+	}
+
+	if watcher == nil {
+		logp.Info("Kubernetes metricset enriching is disabled")
+		return nilEnricher
+	}
+
+	metaConfig := kubernetes.MetaGeneratorConfig{}
+	if err := base.Module().UnpackConfig(&metaConfig); err != nil {
+		logp.Warn("Error initializing Kubernetes metadata enricher: %s", err)
+		return nilEnricher
+	}
+
+	metaGen := kubernetes.NewMetaGeneratorFromConfig(&metaConfig)
+	enricher := buildMetadataEnricher(watcher,
+		// update
+		func(m map[string]common.MapStr, r kubernetes.Resource) {
+			pod := r.(*kubernetes.Pod)
+			meta := metaGen.ResourceMetadata(r)
+			for _, container := range append(pod.GetSpec().GetContainers(), pod.GetSpec().GetInitContainers()...) {
+				id := join(r.GetMetadata().GetNamespace(), r.GetMetadata().GetName(), container.GetName())
+				m[id] = meta
+			}
+		},
+		// delete
+		func(m map[string]common.MapStr, r kubernetes.Resource) {
+			pod := r.(*kubernetes.Pod)
+			for _, container := range append(pod.GetSpec().GetContainers(), pod.GetSpec().GetInitContainers()...) {
+				id := join(r.GetMetadata().GetNamespace(), r.GetMetadata().GetName(), container.GetName())
```

Comment:
You could also have a struct type for the ids, so the map is `map[resourceID]common.MapStr`, and `id := resourceID{r.GetMetadata().GetNamespace(), ...`.
I have had bad experiences in the past building map keys by joining strings :slightly_smiling_face:

Reply:
I would say in general I can do the same mistakes with both approaches, `join` method is syntactically similar to your `resourceID`, in both cases I can pass parameters in the wrong order.